enhancements/update/upgrade-without-registry.md

```diff
@@ -0,0 +1,989 @@
+of each OCP component, and prevent merging if it detects that the offending pull
+policy is used. We should consider adding admission in CI for this.
+
+### Graduation Criteria
+
+The feature will ideally be introduced as `Dev Preview` in OpenShift 4.14.z,
+moved to `Tech Preview` in 4.15 and declared `GA` in 4.16.
+
+#### Dev Preview -> Tech Preview
+
+- Ability to upgrade a single-node clusters using the `bundle.enable: false` mode
+(no bundle, just pinning and pre-loading of the images).
+
+- Availability of the tests that verify the upgrade if single-node clusters.
+
+- Availability of the tests that verify that no OCP component uses the `Always`
+pull policy.
+
+- Obtain positive feedback from at least one customer.
+
+#### Tech Preview -> GA
+
+- Ability to upgrade single-node and multi-node clusters using the
+`bundle.enable` set to `true` or `false`: with a bundle or just pinning and
+pre-loading the images.
+
+- Ability to create bundles using the `oc adm upgrade create bundle` command.
+
+- Availability of the tests that verify the upgrade in single-node and
+multi-node clusters.
+
+- User facing documentation created in
+[https://github.com/openshift/openshift-docs](openshift-docs).
+
+#### Removing a deprecated feature
+
+Not applicable, no feature will be removed.
+
+### Upgrade / Downgrade Strategy
+
+There are no additional considerations for upgrade or downgrade. The same
+considerations that apply to the cluster version operator in general will also
+apply in this case.
+
+### Version Skew Strategy
+
+This feature will only be usable once the cluster version operator and the
+machine config operator have been upgraded to support it. That upgrade will have
+to be done by other means.
+
+For subsequent upgrades we will ensure that the cluster version operator can
+work with both the old and the new version of the machine config operator.
+
+### Operational Aspects of API Extensions
+
+Not applicable, there are no API extensions.
+
+#### Failure Modes
+
+#### Support Procedures
+
+## Implementation History
+
+There is an initial prototype exploring some of the implementation details
+described here in this [https://github.com/jhernand/upgrade-tool](repository).
+
+## Alternatives
+
+The alternative to this is to make a registry server available, either outside
+or inside the cluster.
+
+An external registry server is a well known solution, even for disconnected
+environments, but it is not feasible in most of the target environments.
+
+An internal registry server, running in the cluster itself, is a feasible
+alternative for nodes with multiple clusters. The registry server supported by
+Red Hat is Quay. The disadvantages of Quay is that it requires additional
+resources that are often not available in the target environments.
+
+For single node clusters an internal registry server isn't an alternative
+because it would need to be continuously available during and after the upgrade,
```

Comment:
Take into account that for single-node clusters this registry needs to be running very early. Not sure if a workload managed by the cluster itself will work. An systemd service will certainly work; that is what we use for the installation of the OpenShift appliance. The question is then if we want to support that new piece of software. For the installation it is OK because it only runs once, during the installation. But for upgrades it would need to be a new component that needs itself to be patched and upgraded.

Reply:
We have other containers executed via systemd+podman for the "on prem" cluster types; they're still "managed" though because we generate those systemd units (which fetch via digest) via machineconfig.

Reply:
That is, more or less, the approach that is used for installation in the appliance: a systemd service (created via ignition) that uses podman to run a docker/distribution registry to serve images from a dedicated data partition. But we shut it down after installation, because we don't want to patch and upgrade it. If having this piece of software in every cluster (maybe down, and automatically activated when required via systemd) then it would help. But this new service would need to be owned and maintained by the OCP team. Does that sound reasonable?